src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Runtime/CSharpDetectPreviewFeatureAnalyzer.cs

```diff
@@ -11,6 +12,218 @@ namespace Microsoft.NetCore.CSharp.Analyzers.Runtime
+            {
+                if (typeArgument is GenericNameSyntax innerGenericName)
+                {
+                    if (TryMatchGenericSyntaxNodeWithGivenSymbol(innerGenericName, previewReturnTypeSymbol, out syntaxNode))
+                    {
+                        return true;
+                    }
+                }
+                if (IsIdentifierNameSyntax(typeArgument, previewReturnTypeSymbol))
+                {
+                    syntaxNode = typeArgument;
+                    return true;
+                }
+            }
+
+            syntaxNode = null;
+            return false;
+        }
+
+        protected override SyntaxNode? GetConstraintSyntaxNodeForTypeConstrainedByPreviewTypes(ISymbol typeOrMethodSymbol, ISymbol previewInterfaceConstraintSymbol)
+        {
+            ImmutableArray<SyntaxReference> typeSymbolDeclaringReferences = typeOrMethodSymbol.DeclaringSyntaxReferences;
+
+            foreach (SyntaxReference? syntaxReference in typeSymbolDeclaringReferences)
+            {
+                SyntaxNode typeOrMethodDefinition = syntaxReference.GetSyntax();
+                if (typeOrMethodDefinition is ClassDeclarationSyntax classDeclaration)
+                {
+                    // For ex: class A<T> where T : IFoo, new() // where IFoo is preview
+                    SyntaxList<TypeParameterConstraintClauseSyntax> constraintClauses = classDeclaration.ConstraintClauses;
+                    if (TryGetConstraintClauseNode(constraintClauses, previewInterfaceConstraintSymbol, out SyntaxNode? ret))
+                    {
+                        return ret;
+                    }
+                }
+                else if (typeOrMethodDefinition is MethodDeclarationSyntax methodDeclaration)
+                {
+                    SyntaxList<TypeParameterConstraintClauseSyntax> constraintClauses = methodDeclaration.ConstraintClauses;
+                    if (TryGetConstraintClauseNode(constraintClauses, previewInterfaceConstraintSymbol, out SyntaxNode? ret))
+                    {
+                        return ret;
+                    }
+                }
+            }
+
+            return null;
+        }
+
+        private static bool TryGetConstraintClauseNode(SyntaxList<TypeParameterConstraintClauseSyntax> constraintClauses, ISymbol previewInterfaceConstraintSymbol, [NotNullWhen(true)] out SyntaxNode? syntaxNode)
+        {
+            foreach (TypeParameterConstraintClauseSyntax constraintClause in constraintClauses)
+            {
+                SeparatedSyntaxList<TypeParameterConstraintSyntax> constraints = constraintClause.Constraints;
+                foreach (TypeParameterConstraintSyntax? constraint in constraints)
+                {
+                    if (constraint is TypeConstraintSyntax typeConstraintSyntax)
+                    {
+                        if (typeConstraintSyntax.Type is IdentifierNameSyntax identifier)
```

Comment:
This doesn't seem sufficient. `TypeConstraintSyntax.Type` can be expected to be of other types.

`NullableTypeSyntax` or `AliasQualifiedNameSyntax` for example. (I'm not sure if there are other expected types too)

Reply:
I'm not sure how to do this semantically. But I think it might at least worth extending this to other syntax types. At least `NullableTypeSyntax` which can be very common.